timemachine/cpp/src/context.cu

```diff
@@ -38,6 +52,326 @@ Context::~Context() {
+        if (thermostat->get_temperature() != temperature) {
+            throw std::runtime_error("Local MD temperature didn't match Thermostat's temperature.");
+        }
+    }
+}
+
+std::array<std::vector<double>, 2> Context::multiple_steps_local(
+    const int n_steps,
+    const std::vector<int> &local_idxs,
+    const int store_x_interval,
+    const double radius,
+    const double k,
+    const double temperature,
+    const int seed) {
+    if (store_x_interval <= 0) {
+        throw std::runtime_error("store_x_interval <= 0");
+    }
+    this->_assert_temperature_matches(temperature);
+
+    const int x_buffer_size = n_steps / store_x_interval;
+
+    const int box_buffer_size = x_buffer_size * 3 * 3;
+
+    std::vector<std::shared_ptr<Potential>> initial_potentials(bps_.size());
+
+    for (BoundPotential *pot : bps_) {
+        initial_potentials.push_back(pot->potential);
+    }
+    std::vector<std::shared_ptr<Potential>> flattened_potentials;
+    flatten_potentials(initial_potentials, flattened_potentials);
+
+    std::shared_ptr<Potential> nonbonded_potential;
+
+    // Find the nonbonded potential
+    for (std::shared_ptr<Potential> pot : flattened_potentials) {
+        if (is_nonbonded_potential(pot)) {
+            if (nonbonded_potential) {
+                throw std::runtime_error("found multiple NonbondedAllPairs potentials");
+            }
+            nonbonded_potential = pot;
+        }
+    }
+    if (!nonbonded_potential) {
+        throw std::runtime_error("unable to find a NonbondedAllPairs potential");
+    }
+
+    std::mt19937 rng;
+    rng.seed(seed);
+    std::uniform_int_distribution<unsigned int> random_dist(0, local_idxs.size() - 1);
+
+    // Construct neighborlist to find the inner and outer shell
+    Neighborlist<float> nblist(N_);
+
+    // Store coordinates in host memory as it can be very large
+    std::vector<double> h_x_buffer(x_buffer_size * N_ * 3);
+    // Store boxes on GPU as boxes are a constant size and relatively small
+    std::unique_ptr<DeviceBuffer<double>> d_box_traj(nullptr);
+    if (box_buffer_size > 0) {
+        d_box_traj.reset(new DeviceBuffer<double>(box_buffer_size));
+    }
+
+    const size_t tpb = warp_size;
+
+    DeviceBuffer<unsigned int> d_shell_idxs_inner(N_);
+
+    DeviceBuffer<unsigned int> d_row_idxs(N_);
+    // d_col indices used both for column indices for neighborlist as well as outer shell.
+    DeviceBuffer<unsigned int> d_col_idxs(N_);
+
+    // Pinned memory for getting lengths of indice arrays
+    PinnedHostBuffer<int> p_num_selected(1);
+    DeviceBuffer<int> num_selected_buffer(1);
+    LessThan select_op(N_);
+
+    std::size_t temp_storage_bytes = 0;
+    cub::DevicePartition::If(
+        nullptr, temp_storage_bytes, d_shell_idxs_inner.data, d_row_idxs.data, num_selected_buffer.data, N_, select_op);
+    DeviceBuffer<char> d_temp_storage_buffer(temp_storage_bytes);
```

Comment:
Not quite following this -- is it worth a line comment why this temporary storage buffer is of bytes rather than ints / uints?

(Will it only ever contain idxs? Or will it also be used for something else?)

Reply:
Addressed by comment in https://github.com/proteneer/timemachine/pull/915/commits/57ef840f794f31cee087c2686d14974f55c787e1